lib/pure/endians.nim

```diff
@@ -98,16 +124,734 @@ else:
+
+  proc bigEndian32*(outp, inp: pointer) {.inline, deprecated:
+      "Deprecated since v1.1.0; use `fromBytesXX/toBytesXX` instead".} =
+    swapEndian32(outp, inp)
+
+  proc bigEndian16*(outp, inp: pointer) {.inline, deprecated:
+      "Deprecated since v1.1.0; use `fromBytesXX/toBytesXX` instead".} =
+    swapEndian16(outp, inp)
+
+
+# New API starts here
+
+func slowSwap16(a: uint16): uint16 {.inline.} =
+  result = (a shl 8) or (a shr 8)
+
+func slowSwap32(a: uint32): uint32 {.inline.} =
+  var a = ((a shl  8) and 0xff00ff00'u32) or
+          ((a shr  8) and 0x00ff00ff'u32)
+  result = (a shl 16) or (a shr 16)
+
+func slowSwap64(a: uint64): uint64 {.inline.} =
+  var a = ((a shl 8) and 0xff00ff00ff00ff00'u64) or
+          ((a shr 8) and 0x00ff00ff00ff00ff'u64)
+  a = ((a shl 16) and 0xffff0000ffff0000'u64) or
+      ((a shr 16) and 0x0000ffff0000ffff'u64)
+  result = (a shl 32) or (a shr 32)
+
+when useBuiltinSwap:
+  template swap16(a: uint16): uint16 = builtinSwap16(a)
+  template swap32(a: uint32): uint32 = builtinSwap32(a)
+  template swap64(a: uint64): uint64 = builtinSwap64(a)
 else:
-  proc littleEndian64*(outp, inp: pointer) {.inline.} = copyMem(outp, inp, 8)
-  proc littleEndian32*(outp, inp: pointer) {.inline.} = copyMem(outp, inp, 4)
-  proc littleEndian16*(outp, inp: pointer){.inline.} = copyMem(outp, inp, 2)
-  proc bigEndian64*(outp, inp: pointer) {.inline.} = swapEndian64(outp, inp)
-  proc bigEndian32*(outp, inp: pointer) {.inline.} = swapEndian32(outp, inp)
-  proc bigEndian16*(outp, inp: pointer) {.inline.} = swapEndian16(outp, inp)
+  template swap16(a: uint16): uint16 = slowSwap16(a)
+  template swap32(a: uint32): uint32 = slowSwap32(a)
+  template swap64(a: uint64): uint64 = slowSwap64(a)
+
+
+# TODO had to remove {.since: (1,1).} because it broke `koch runCI`
+func swapEndian*[T: SomeNumber](value: T): T {.inline.} =
+  ## Swaps the byte order of a number.
+  runnableExamples:
+    doAssert swapEndian(0xdeadbeef'i32) == 0xefbeadde'i32
+    doAssert swapEndian(0xcafe'u16) == 0xfeca'u16
+    doAssert swapEndian(0x42'u8) == 0x42'u8
+
+  when sizeof(T) == 1:
+    value
+  elif sizeof(T) == 2:
+    when nimvm: cast[T](slowSwap16(cast[uint16](value)))
+    else:       cast[T](    swap16(cast[uint16](value)))
+  elif sizeof(T) == 4:
+    when nimvm: cast[T](slowSwap32(cast[uint32](value)))
+    else:       cast[T](    swap32(cast[uint32](value)))
+  elif sizeof(T) == 8:
+    when nimvm: cast[T](slowSwap64(cast[uint64](value)))
+    else:       cast[T](    swap64(cast[uint64](value)))
+
+
+func swapEndian*(T: typedesc[SomeNumber],
+                 buf: pointer) {.inline, since: (1,1).} =
+  ## Swaps the byte order of a number at a memory location in-place. The type
+  ## of the number must be passed in as an argument.
+  runnableExamples:
+    var i = 0xdeadbeef'i32
+    swapEndian(int32, i.addr)
+    doAssert i == 0xefbeadde'i32
+
+  var value = cast[ptr T](buf)[]
+  cast[ptr T](buf)[] = swapEndian(value)
```

Comment:
here's the alignment bug, from what I can tell: 

if `buf` points to an address that is not aligned for T, it's not valid to write a T-sized value to it. unaligned writes often happen when you're putting things in network buffers that are packed - the very use case you're trying to be convenient for.

this is one way in which this API is unsafe - the other is obviously buffer overruns - I'd prefer to see that the std library is kept memory-safe as far as possible.